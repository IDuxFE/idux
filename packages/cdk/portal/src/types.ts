/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { IxInnerPropTypes, IxPublicPropTypes } from '@idux/cdk/utils'
import type { DefineComponent } from 'vue'

import { IxPropTypes } from '@idux/cdk/utils'

export const portalTargetDef = IxPropTypes.oneOfType([
  String,
  HTMLElement,
  IxPropTypes.func<() => string | HTMLElement>(),
])

export const portalProps = {
  disabled: IxPropTypes.bool,
  target: IxPropTypes.oneOfType([String, HTMLElement, IxPropTypes.func<() => string | HTMLElement>()]).isRequired,
  load: IxPropTypes.bool.def(true),
}

export type PortalProps = IxInnerPropTypes<typeof portalProps>
export type PortalPublicProps = IxPublicPropTypes<typeof portalProps>
export type PortalComponent = DefineComponent<PortalPublicProps>
export type PortalInstance = InstanceType<DefineComponent<PortalProps>>

export type PortalTargetType = string | HTMLElement | (() => string | HTMLElement)
