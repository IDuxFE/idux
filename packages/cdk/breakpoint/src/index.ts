export * from './observe'
export * from './breakpoints'
export * from './mediaMatcher'
