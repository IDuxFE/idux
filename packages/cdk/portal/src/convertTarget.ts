/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { PortalTargetType } from './types'

import { isFunction, isString } from 'lodash-es'

function tryFindElement(selector: string) {
  const classSelector = `.${selector}`
  return document.querySelector<HTMLElement>(selector) ?? document.querySelector<HTMLElement>(classSelector)
}

export function convertTarget(target: PortalTargetType): HTMLElement {
  const temp = isFunction(target) ? target() : target

  if (!isString(temp)) {
    return temp
  }

  let element = tryFindElement(temp)
  if (!element) {
    element = document.createElement('div')
    element.classList.add(temp)
    document.body.appendChild(element)
  }

  return element
}
