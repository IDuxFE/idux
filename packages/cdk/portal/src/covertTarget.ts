/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { PortalTargetType } from './types'

import { isFunction, isString } from 'lodash-es'

const targetHashmap: Record<string, HTMLElement> = {}

export function covertTarget(target: PortalTargetType): HTMLElement {
  const temp = isFunction(target) ? target() : target

  if (!isString(temp)) {
    return temp
  }

  if (targetHashmap[temp]) {
    return targetHashmap[temp]
  }

  let element = document.querySelector<HTMLElement>(temp)
  if (!element) {
    element = document.createElement('div')
    element.classList.add(temp)
    document.body.appendChild(element)
  }

  targetHashmap[temp] = element

  return element
}
