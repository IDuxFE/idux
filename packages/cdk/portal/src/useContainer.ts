/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import { isString } from 'lodash-es'

const containerHashmap: Record<string, HTMLElement> = {}

export function useContainer(target: string | HTMLElement): HTMLElement {
  if (!isString(target)) {
    return target
  }
  if (containerHashmap[target]) {
    return containerHashmap[target]
  }
  const container = document.createElement('div')
  container.classList.add(target)
  document.body.appendChild(container)
  containerHashmap[target] = container
  return container
}
