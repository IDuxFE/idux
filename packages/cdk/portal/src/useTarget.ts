/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import { isString } from 'lodash-es'

const targetHashmap: Record<string, HTMLElement> = {}

export function useTarget(target: string | HTMLElement): HTMLElement {
  if (!isString(target)) {
    return target
  }
  if (targetHashmap[target]) {
    return targetHashmap[target]
  }
  const Target = document.createElement('div')
  Target.classList.add(target)
  document.body.appendChild(Target)
  targetHashmap[target] = Target
  return Target
}
