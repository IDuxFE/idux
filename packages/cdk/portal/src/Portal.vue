<template>
  <teleport :disabled="disabled" :to="to">
    <slot />
  </teleport>
</template>

<script lang="ts">
import type { ComputedRef } from 'vue'
import type { PortalProps } from './types'

import { computed, defineComponent } from 'vue'
import { PropTypes } from '@idux/cdk/utils'

import { useContainer } from './useContainer'

export default defineComponent({
  name: 'IxPortal',
  props: {
    disabled: PropTypes.bool,
    target: PropTypes.oneOfType([PropTypes.string, PropTypes.instanceOf(HTMLElement)]).isRequired,
  },
  setup(props: PortalProps) {
    const to = useTarget(props)

    return { to }
  },
})

function useTarget(props: PortalProps): ComputedRef<HTMLElement> {
  return computed(() => useContainer(props.target))
}
</script>
