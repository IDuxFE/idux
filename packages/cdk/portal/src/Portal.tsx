/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import { Teleport, computed, defineComponent, ref, watch } from 'vue'

import { portalProps } from './types'
import { useTarget } from './useTarget'

export default defineComponent({
  name: 'IxPortal',
  props: portalProps,
  setup(props) {
    const loaded = ref(props.load)
    watch(
      () => props.load,
      load => {
        if (!loaded.value) {
          loaded.value = load
        }
      },
    )
    const to = computed(() => loaded.value && useTarget(props.target))
    return { to }
  },
  render() {
    const { to, disabled, $slots } = this
    if (!to) {
      return null
    }

    return (
      <Teleport to={to} disabled={disabled}>
        {$slots.default?.()}
      </Teleport>
    )
  },
})
