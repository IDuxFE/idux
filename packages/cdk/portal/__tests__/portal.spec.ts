import { MountingOptions, mount } from '@vue/test-utils'

import CdkPortal from '../src/Portal'
import { PortalProps } from '../src/types'

describe('Portal', () => {
  const PortalMount = (options?: MountingOptions<Partial<PortalProps>>) =>
    mount(CdkPortal, {
      slots: { default: `<div id='content'>content</div>` },
      ...(options as MountingOptions<PortalProps>),
      attachTo: 'body',
    })

  afterEach(() => {
    document.body.childNodes.forEach(node => node.remove())
  })

  test('render work', () => {
    const wrapper = PortalMount({ props: { target: '.render-container' } })
    expect(document.body.querySelector('.render-container')!.querySelector('#content')).toBeDefined()

    expect(() => {
      wrapper.vm.$forceUpdate()
      wrapper.unmount()
    }).not.toThrow()
  })

  test('disabled work', async () => {
    const wrapper = PortalMount({ props: { target: '.disabled-container', disabled: true } })
    const container = document.body.querySelector('.disabled-container')!
    expect(container.querySelector('#content')).toBeNull()

    await wrapper.setProps({ disabled: false })

    expect(container.querySelector('#content')).not.toBeNull()
  })

  test('target work', async () => {
    const wrapper = PortalMount({ props: { target: '.target-container1' } })
    const container1 = document.body.querySelector('.target-container1')!
    expect(container1.querySelector('#content')).not.toBeNull()

    const container2 = document.createElement('div')
    container2.classList.add('target-container2')
    document.body.appendChild(container2)

    await wrapper.setProps({ target: container2 })
    expect(container1.querySelector('#content')).toBeNull()
    expect(container2.querySelector('#content')).not.toBeNull()

    const container3 = document.createElement('div')
    container3.classList.add('target-container3')
    document.body.appendChild(container3)

    await wrapper.setProps({ target: '.target-container3' })
    expect(container1.querySelector('#content')).toBeNull()
    expect(container2.querySelector('#content')).toBeNull()
    expect(container3.querySelector('#content')).not.toBeNull()

    const container4 = document.createElement('div')
    container4.id = 'target-container4'
    document.body.appendChild(container4)

    await wrapper.setProps({ target: '#target-container4' })
    expect(container1.querySelector('#content')).toBeNull()
    expect(container2.querySelector('#content')).toBeNull()
    expect(container3.querySelector('#content')).toBeNull()
    expect(container4.querySelector('#content')).not.toBeNull()
  })

  test('load work', async () => {
    const wrapper = PortalMount({ props: { target: '.load-container', load: false } })
    expect(document.body.querySelector('.load-container')).toBeNull()

    await wrapper.setProps({ load: true })

    expect(document.body.querySelector('.load-container')).not.toBeNull()
  })
})
