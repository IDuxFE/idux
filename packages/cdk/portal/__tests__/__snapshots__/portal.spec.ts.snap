// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Portal.vue disabled work 1`] = `
"<!--teleport start-->
<!--teleport end-->"
`;

exports[`Portal.vue render work 1`] = `
"<!--teleport start-->
<!--teleport end-->"
`;

exports[`Portal.vue target work 1`] = `
"<!--teleport start-->
<!--teleport end-->"
`;

exports[`Portal.vue target work 2`] = `
"<!--teleport start-->
<!--teleport end-->"
`;
